src/coordinator/executor/transform/cache.go

```diff
@@ -70,3 +71,92 @@ func (c *BlockCache) Get(key parser.NodeID) (block.Block, bool) {
 	b, ok := c.blocks[key]
 	return b, ok
 }
+
+type cacheTime int64
+
+func fromTime(t time.Time) cacheTime {
+	return cacheTime(t.UnixNano())
+}
+
+func (c cacheTime) toTime() time.Time {
+	return time.Unix(0, int64(c))
+}
+
+// TimeCache is used to cache blocks over time. It also keeps track of blocks which have been processed
+type TimeCache struct {
+	blocks          map[cacheTime]block.Block
+	processedBlocks map[cacheTime]bool
+	mu              sync.Mutex
+}
+
+// NewTimeCache creates a new TimeCache
+func NewTimeCache() *TimeCache {
+	return &TimeCache{
+		blocks:          make(map[cacheTime]block.Block),
+		processedBlocks: make(map[cacheTime]bool),
+	}
+}
+
+// Add the block to the cache, errors out if block already exists
+func (c *TimeCache) Add(key time.Time, b block.Block) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	_, ok := c.blocks[fromTime(key)]
+	if ok {
+		return errors.New("block already exists")
+	}
+
+	c.blocks[fromTime(key)] = b
+	return nil
+}
+
+// Remove the block from the cache
+func (c *TimeCache) Remove(key time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.blocks, fromTime(key))
+}
+
+// Get the block from the cache
+func (c *TimeCache) Get(key time.Time) (block.Block, bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
Rather than incurring the cost of defer here, unlock after retrieving the block manually